python/ray/rllib/optimizers/aso_learner.py

```diff
@@ -43,18 +50,47 @@ def __init__(self, local_worker, minibatch_buffer_size, num_sgd_iter,
         self.stats = {}
         self.stopped = False
 
+
     def run(self):
         while not self.stopped:
             self.step()
 
     def step(self):
         with self.queue_timer:
-            batch, _ = self.minibatch_buffer.get()
+            # old_worker not None means that APPO agent is being used (the only agent that uses old_worker)
+            if self.old_worker:
+                batch, _ = self.minibatch_buffer.get()
+            else:
+                batch = self.inqueue.get()
+
+        if self.old_worker and "old_policy_behaviour_logits" not in batch:
+            batch["old_policy_behaviour_logits"] = self.old_worker.policy_map['default_policy'].compute_actions(
+                batch["obs"],prev_action_batch=batch["prev_actions"],
+                prev_reward_batch=batch["prev_rewards"])[2]['behaviour_logits']
 
         with self.grad_timer:
             fetches = self.local_worker.learn_on_batch(batch)
             self.weights_updated = True
             self.stats = get_learner_stats(fetches)
 
+        if self.old_worker:
```

Comment:
I don't like all of these "if" cases - can we separate this into a separate component? Like 
```class ImportanceSamplingBiaser
def __init__(self, local worker):

def update_if_needed(self):
```

and all of this code is basically encapsulated into that. Honestly, can't that actually just be a mixin?

Reply:
Good idea! I am not sure though if mixin would be the easiest solution since both old_worker and the kl list tracking (for updating kl coeffs) depend on self.counter, which is incremented in the learner thread  (every call to step).

Reply:
Copied from slack: Looked into adding an extra mixin for importance sampling. Don't think it is possible since, for every time train is called for APPOTrainer, couple iterations of LearnerThread's step is called in the background. Old worker (pi_old) is very sensitive to  how many times LearnerThread is called and obtaining "counter" from such a mixin will not give the exact "counter" that is needed to update LearnerThread.
Are you fine if an additional class is made in aso_learner.py to wrap around the old_worker?

Reply:
To work around this, I made an extra class called APPOTracker in aso_learner.py. Lmk if it makes the LearnerThread class easier to read.